obsplus/structures/grid.py

```diff
@@ -0,0 +1,1323 @@
+    points[zout] = points[zcol]
+
+    # Deal with a callable conversion
+    if isinstance(conversion, Callable):
+        try:
+            if conversion_kwargs is not None:
+                points = conversion(points, **conversion_kwargs)
+            else:
+                points = conversion(points)
+        except Exception as e:
+            raise RuntimeError(f"Conversion callable raised an exception: {e}")
+        if not isinstance(points, pd.DataFrame):
+            raise ValueError("Callable conversion must return a pandas DataFrame")
+        return points
+
+    # Do some checking to make sure the conversion is kosher
+    if not isinstance(conversion, Iterable):
+        raise TypeError(f"conversion shoud be a list of transformations")
+    try:
+        if not conversion[0][0].lower() in valid_keys:
+            raise TypeError(f"conversion should be a list of transformations")
+    except AttributeError:
+        raise TypeError(f"conversion should be a list of transformations")
+
+    # Go through each stage of the conversion
+    for stage in conversion:
+        # Make sure the stage is valid
+        key = stage[0].lower()
+        value = stage[1]
+        if key not in valid_keys:
+            raise ValueError(f"Invalid conversion operation: {key}")
+        elif not key == "project":
+            try:
+                value = float(value)
+            except ValueError:
+                raise TypeError(f"value must be a float for key: {key}, {value}")
+        # Do the conversion
+        if key == "scale_x":
+            points[xout] = points[xout] * value
+        elif key == "scale_y":
+            points[yout] = points[yout] * value
+        elif key == "scale_z":
+            points[zout] = points[zout] * value
+        elif key == "translate_x":
+            points[xout] = points[xout] + value
+        elif key == "translate_y":
+            points[yout] = points[yout] + value
+        elif key == "translate_z":
+            points[zout] = points[zout] + value
+        elif key == "rotate_xy":
+            points[xout], points[yout] = rotate_point(points[xout], points[yout], value)
+        elif key == "rotate_xz":
+            points[xout], points[zout] = rotate_point(points[xout], points[zout], value)
+        elif key == "rotate_yz":
+            points[yout], points[zout] = rotate_point(points[yout], points[zout], value)
+        elif key == "project":
+            try:
+                try:
+                    fro = pyproj.Proj(str(value["from"]))
+                except RuntimeError:
+                    msg = "%s is not a valid projection" % value["from"]
+                    _log_raise(msg, ValueError)
+                try:
+                    to = pyproj.Proj(str(value["to"]))
+                except RuntimeError:
+                    msg = "%s is not a valid projection" % value["to"]
+                    _log_raise(msg, ValueError)
+                #                for i, point in points.iterrows():
```

Comment:
Can these be deleted?

Reply:
Indeed. I'm amazed that I missed that one.